crucible-go/src/Lang/Crucible/Go/TransUtil.hs

```diff
@@ -0,0 +1,756 @@
+
+import           Data.Parameterized.Context as Ctx
+import           Data.Parameterized.Some (Some(..))
+import           Data.Parameterized.TraversableFC
+
+import           Debug.Trace (trace)
+
+import           What4.Utils.StringLiteral
+
+import qualified Lang.Crucible.CFG.Core as C
+import qualified Lang.Crucible.CFG.Expr as C
+import qualified Lang.Crucible.CFG.Generator as Gen
+import           Lang.Crucible.Syntax
+import           Lang.Crucible.Types
+
+import           Language.Go.AST
+import           Language.Go.Rec
+import           Language.Go.Types
+import           Lang.Crucible.Go.Encodings
+import           Lang.Crucible.Go.Types
+
+-- | Use this function (or one of its variants) to assert that an
+-- expression has a certain type. Unit values may be coerced to
+-- Nothing values of a Maybe type. Floats may be cast (usually from
+-- float literals being double by default but allowed to implicitly
+-- cast to single precision). It may be useful in the future to do a
+-- type inference pass, inferring the actual eventual types of
+-- "untyped constants", to obviate the need for this coercion
+-- stuff. It seems that the Go typechecker is doing this for us for
+-- integer constants but not floats or nil.
+asTypeEither :: TypeRepr b
+             -> Gen.Expr Go s a
+             -> Either (Some TypeRepr, Some (Gen.Expr Go s)) (Gen.Expr Go s b)
+asTypeEither repr e =
+  -- trace ("coercing " ++ show (exprType e) ++ " to " ++ show repr) $
+  case testEquality (exprType e) repr of
+    Just Refl -> Right e
+    Nothing ->
+      -- When the types are not equal, try to coerce the value.
+      case (e, exprType e, repr) of
+        -- Send unit to Nothing.
+        (Gen.App C.EmptyApp, _repr, MaybeRepr repr') ->
+          Right $ Gen.App $ C.NothingValue repr'
+        -- Coerce floats (usually double to float)
+        (_e, FloatRepr _fi, FloatRepr fi) ->
+          Right $ Gen.App $ C.FloatCast fi C.RNE e
+        _ -> Left (Some repr, Some e)
+
+asTypeMaybe :: TypeRepr b -> Gen.Expr Go s a -> Maybe (Gen.Expr Go s b)
+asTypeMaybe repr e = case asTypeEither repr e of
+  Left _x -> Nothing
+  Right e' -> Just e'
+
+asType :: TypeRepr b -> Gen.Expr Go s a -> Gen.Expr Go s b
+asType repr e =
+  case asTypeMaybe repr e of
+    Just e' -> e'
+    Nothing -> error $ "asType fail: expression " ++ show e ++ " of type "
+               ++ show (exprType e) ++ " incompatible with type " ++ show repr
+
+-- | CPS version of asType
+asType' :: TypeRepr b -> Gen.Expr Go s a -> (Gen.Expr Go s b -> c) -> c
+asType' repr e k = k $ asType repr e
+
+-- | asTypeEither lifted to assignments.
+asTypesEither :: CtxRepr ctx'
+              -> Assignment (Gen.Expr Go s) ctx
+              -> Either (Some TypeRepr, Some (Gen.Expr Go s))
+              (Assignment (Gen.Expr Go s) ctx')
+asTypesEither Empty Empty = Right Empty
+asTypesEither (reprs :> repr) (es :> e) =
+  pure (:>) <*> asTypesEither reprs es <*> asTypeEither repr e
```

Comment:
I think that is the same as `(:>) <$> asTypesEither reprs es <*> asTypeEither repr e`

Reply:
Fixed.